### PR TITLE
Xdai eip1559 - implementation draft

### PR DESCRIPTION
**Pull request by MarekM25**
Guys, with this PR I wanted only to get your feedback. I don't have automatic tests and don't perform manual testing. It is not ready to merge!

I would like to discuss:
- [ ] adding MinGasPremium to MiningConfig and maybe MinGasPremiumTxFilter? Do we want something like that?

ToDo list:
- [x] prepare draft of implementation
- [ ] add automatic tests
- [ ] perform manual testing
- [ ] prepare performance tests

### CODE REVIEW

File: src/Nethermind/Nethermind.Blockchain.Test/TxPools/Collections/SortedPoolTests.cs [hunk 1]
```diff
@@ -31,13 +35,20 @@ public class SortedPoolTests
...
         
         [SetUp]
         public void Setup()
         {
+            ISpecProvider specProvider = Substitute.For<ISpecProvider>();
+            IBlockTree blockTree = Substitute.For<IBlockTree>();
+            Block block =  Build.A.Block.WithNumber(0).TestObject;
+            blockTree.Head.Returns(block);
+            specProvider.GetSpec(Arg.Any<long>()).Returns(new ReleaseSpec() {IsEip1559Enabled = false});
+            ITransactionComparerProvider transactionComparerProvider = new TransactionComparerProvider(specProvider, blockTree);
```
Comment (tkstanczak) [inline src/Nethermind/Nethermind.Blockchain.Test/TxPools/Collections/SortedPoolTests.cs:50]: confusing -> what does GetDefaultComparer do?

Comment (tkstanczak) [reply src/Nethermind/Nethermind.Blockchain.Test/TxPools/Collections/SortedPoolTests.cs:50]: why not simple new TxComparer(specProvider, blockTree) here
it seems that we create a provider, use it once and then discard?

File: src/Nethermind/Nethermind.Blockchain/Comparers/TransactionComparerProvider.cs [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+        {
+            _specProvider = specProvider;
+            _blockTree = blockTree;
+            _gasPriceTxComparerByBaseFee = new GasPriceTxComparerByBaseFee(specProvider);
+
+        }
+
+        public IComparer<Transaction> GetDefaultComparer()
+        {
+            if (_defaultComparer == null)
```
Comment (tkstanczak) [inline src/Nethermind/Nethermind.Blockchain/Comparers/TransactionComparerProvider.cs:43]: ok, now I see

File: src/Nethermind/Nethermind.Blockchain/Comparers/TransactionComparerProvider.cs [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+                    .ThenBy(CompareTxByTimestamp.Instance)
+                    .ThenBy(CompareTxByPoolIndex.Instance)
+                    .ThenBy(CompareTxByGasLimit.Instance);
+            }
+
+            return _defaultComparer;
+        }
+
+        public IComparer<Transaction> GetDefaultProducerComparer(IPreparingBlockContextService preparingBlockContextService)
+        {
```
Comment (tkstanczak) [inline src/Nethermind/Nethermind.Blockchain/Comparers/TransactionComparerProvider.cs:56]: nice, I think it is nice

File: src/Nethermind/Nethermind.Blockchain/Producers/TxPoolTxSource.cs [hunk 4]
```diff
@@ -160,11 +167,20 @@ bool HasEnoughFounds(IDictionary<Address, UInt256> balances, Transaction transac
                     continue;
                 }
 
-                if (!HasEnoughFounds(remainingBalance, tx))
+                // ToDo
+                // if (!HasCorrectFeeCap(tx, isEip1559Enabled, baseFee))
```
Comment (tkstanczak) [inline src/Nethermind/Nethermind.Blockchain/Producers/TxPoolTxSource.cs:171]: is this TODO still a TODO?

Comment (tkstanczak) [reply src/Nethermind/Nethermind.Blockchain/Producers/TxPoolTxSource.cs:171]: use TODO: then it is recognized by the IDE

Comment (MarekM25) [reply src/Nethermind/Nethermind.Blockchain/Producers/TxPoolTxSource.cs:171]: cosmetic - I didn't remove ToDo :)

File: src/Nethermind/Nethermind.Consensus.AuRa/Transactions/TxGasPrice1559Sender.cs [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+        }
+
+        private UInt256 CurrentMinGasPremium() =>
+            _txPool.GetPendingTransactions()
+                .Select(t => t.GasPrice)
+                .Where(g => g > UInt256.Zero)
+                .DefaultIfEmpty(TxGasPriceSenderConstants.DefaultGasPrice)
+                .Min();
+        
+        private UInt256 CurrentMinFeeCap() =>
```
Comment (tkstanczak) [inline src/Nethermind/Nethermind.Consensus.AuRa/Transactions/TxGasPrice1559Sender.cs:60]: remove LINQ from here -> performance killes and SendTransaction may be used a lot by some users

File: src/Nethermind/Nethermind.Consensus.AuRa/Transactions/TxGasPrice1559Sender.cs [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+            UInt256 minGasPremium =  CurrentMinGasPremium();
+            UInt256 minFeeCap =  CurrentMinFeeCap();
+            UInt256 txGasPrice = minGasPremium * _percentDelta / 100;
+            tx.GasPrice = UInt256.Max(txGasPrice, _miningConfig.MinGasPrice);
+            tx.DecodedFeeCap = UInt256.Max(minFeeCap, _miningConfig.MinGasPrice);
+            return _txSender.SendTransaction(tx, txHandlingOptions);
+        }
+
+        private UInt256 CurrentMinGasPremium() =>
+            _txPool.GetPendingTransactions()
```
Comment (tkstanczak) [inline src/Nethermind/Nethermind.Consensus.AuRa/Transactions/TxGasPrice1559Sender.cs:54]: remove LINQ from here -> performance killer

File: src/Nethermind/Nethermind.Consensus.AuRa/Transactions/TxGasPrice1559Sender.cs [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+        {
+            _txSender = txSender ?? throw new ArgumentNullException(nameof(txSender));
+            _txPool = txPool ?? throw new ArgumentNullException(nameof(txPool));
+            _miningConfig = miningConfig ?? throw new ArgumentNullException(nameof(miningConfig));
+            _percentDelta = percentDelta;
+        }
+
+        public ValueTask<Keccak> SendTransaction(Transaction tx, TxHandlingOptions txHandlingOptions)
+        {
+            UInt256 minGasPremium =  CurrentMinGasPremium();
```
Comment (tkstanczak) [inline src/Nethermind/Nethermind.Consensus.AuRa/Transactions/TxGasPrice1559Sender.cs:45]: you call GetPendingTransactions twice here -> you will have inconsistent results and also big performance hit -> call it once then, use it further

File: src/Nethermind/Nethermind.Consensus.AuRa/Transactions/TxGasPrice1559Sender.cs (hunk 5 above)
Comment (tkstanczak) [reply src/Nethermind/Nethermind.Consensus.AuRa/Transactions/TxGasPrice1559Sender.cs:60]: :O :O

Comment (tkstanczak) [reply src/Nethermind/Nethermind.Consensus.AuRa/Transactions/TxGasPrice1559Sender.cs:60]: I think I let it in last time when I realized it is in AuRa namespace

**Pull request closed without merging**